### PR TITLE
Optimize PReLU (float32) and enable PReLU BFloat16 support in CPU path

### PR DESCRIPTION
**Pull request by yanbing-j**
In this PR, we try to optimize PReLU op in CPU path, and enable BFloat16 support based on the optimized PReLU.

The original implementation uses parallel_for to accelerate operation speed, but vectorization is not used. It can be optimized by using TensorIterator, both including parallelization and vectorization.

The difference between PReLU and other activation function ops, is that PReLU supports a learnable parameter `weight`. When called without arguments, nn.PReLU() uses a single parameter `weight` across all input channels. If called with nn.PReLU(nChannels), a separate `weight` is used for each input channel. So we cannot simply use TensorIterator because `weight` is different for each input channel.

In order to use TensorIterator, `weight` should be broadcasted to `input` shape. And with vectorization and parallel_for, this implementation is much faster than the original one. Another advantage is, don't need to separate `share weights` and `multiple weights` in implementation.

We test the performance between the PReLU implementation of public Pytorch and the optimized PReLU in this PR, including fp32/bf16, forward/backward, share weights/multiple weights configurations. bf16 in public Pytorch directly reuses `Vectorized<scalar_t>` for `BFloat16`.

Share weights:
image

image

Multiple weights:
image

image

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

**Comment by yanbing-j** [comment]
@jbschlosser Please review this PR. Thanks.

**Comment by xsacha** [comment]
Is the performance of this PR similar to that of PR #60427 ?

**Comment by yanbing-j** [comment]
> Is the performance of this PR similar to that of PR #60427 ?

The performance of multi weights of this PR is similar to that of PR #60427. This PR performs similar with small inputs of share weights, and performs better with large inputs of share weights than that of PR #60427.

**Comment by yanbing-j** [comment]
@VitalyFedyunin Please help review this PR. Thanks.

**Comment by mruberry** [comment]
ping @jbschlosser

**Comment by yanbing-j** [comment]
@VitalyFedyunin Hi, could you please help review this PR? Thanks.

**Comment by yanbing-j** [comment]
Hi @VitalyFedyunin , could you please help review this PR? Customer cares about this usage.

**Comment by VitalyFedyunin** [comment]
I believe @[frank-wei](https://github.com/frank-wei) is looking at this PR.

**Review by frank-wei — APPROVED** [review]
LGTM!

**Comment by frank-wei** [comment]
Could you help check the failure?
test_nn_PReLU_scalar (test_jit.TestJitGeneratedModule) ... python3.8: caffe2/c10/util/SmallVector.h:338: c10::SmallVectorTemplateCommon::reference c10::SmallVectorTemplateCommon<long>::operator[](c10::SmallVectorTemplateCommon::size_type) [T = long]: Assertion `idx < size()' failed.
test_nn_PReLU_no_batch_dim (test_jit.TestJitGeneratedModule) ... python3.8: caffe2/c10/util/SmallVector.h:338: c10::SmallVectorTemplateCommon::reference c10::SmallVectorTemplateCommon<long>::operator[](c10::SmallVectorTemplateCommon::size_type) [T = long]: Assertion `idx < size()' failed.
test_custom_rules_ints_nn_functional_prelu_cpu_bfloat16 (jit.test_dtype_analysis.TestDtypeCustomRulesCPU) ... /data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/jit#binary,link-tree/torch/testing/_internal/common_methods_invocations.py:1780: DeprecationWarning: an integer is required (got type float). Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
 weight_tensor = torch.tensor(weight, device=device, dtype=dtype, requires_grad=requires_grad)
python3.8: caffe2/c10/util/SmallVector.h:338: c10::SmallVectorTemplateCommon::reference c10::SmallVectorTemplateCommon<long>::operator[](c10::SmallVectorTemplateCommon::size_type) [T = long]: Assertion `idx < size()' failed.

**Comment by yanbing-j** [comment]
> Could you help check the failure? test_nn_PReLU_scalar (test_jit.TestJitGeneratedModule) ... python3.8: caffe2/c10/util/SmallVector.h:338: c10::SmallVectorTemplateCommon::reference c10::SmallVectorTemplateCommon::operator [T = long]: Assertion `idx < size()' failed. test_nn_PReLU_no_batch_dim (test_jit.TestJitGeneratedModule) ... python3.8: caffe2/c10/util/SmallVector.h:338: c10::SmallVectorTemplateCommon::reference c10::SmallVectorTemplateCommon<long>::operator[](c10::SmallVectorTemplateCommon::size_type) [T = long]: Assertion `idx < size()' failed. test_custom_rules_ints_nn_functional_prelu_cpu_bfloat16 (jit.test_dtype_analysis.TestDtypeCustomRulesCPU) ... /data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/jit#binary,link-tree/torch/testing/_internal/common_methods_invocations.py:1780: DeprecationWarning: an integer is required (got type float). Implicit conversion to integers using **int** is deprecated, and may be removed in a future version of Python. weight_tensor = torch.tensor(weight, device=device, dtype=dtype, requires_grad=requires_grad) python3.8: caffe2/c10/util/SmallVector.h:338: c10::SmallVectorTemplateCommon::reference c10::SmallVectorTemplateCommon::operator [T = long]: Assertion `idx < size()' failed.

Hi @frank-wei , I cannot reproduce the failure of `test_nn_PReLU_scalar`, `test_nn_PReLU_no_batch_dim` and `test_custom_rules_ints_nn_functional_prelu_cpu_bfloat16` in python 3.8 environment. Could you please provide more details about this? Thanks!

**Comment by yanbing-j** [comment]
Hi @frank-wei , could you please provide the machine info (avx2 or avx512), and gcc version as well?

**Comment by frank-wei** [comment]
> Hi @frank-wei , could you please provide the machine info (avx2 or avx512), and gcc version as well?

It is run in the remote machine. I will need to run on my local again to reprod this error and provide with you the machine info later.

**Comment by yanbing-j** [comment]
> > Hi @frank-wei , could you please provide the machine info (avx2 or avx512), and gcc version as well?
> 
> @yanbing-j It was run in the remote machine. I will need to run on my local again to reprod this error and provide with you the machine info later.

@frank-wei Thank you so much!

**Comment by frank-wei** [comment]
> > > Hi @frank-wei , could you please provide the machine info (avx2 or avx512), and gcc version as well?
> > 
> > 
> > @yanbing-j It was run in the remote machine. I will need to run on my local again to reprod this error and provide with you the machine info later.
> 
> @frank-wei Thank you so much!
@yanbing-j ， here is my local machine info. The test error is still there.
processor       : 31
vendor_id       : GenuineIntel
cpu family      : 6
model           : 85
model name      : Intel Xeon Processor (Skylake)
stepping        : 4
microcode       : 0x1
cpu MHz         : 1596.324
cache size      : 16384 KB
physical id     : 31
siblings        : 1
core id         : 0
cpu cores       : 1
apicid          : 31
initial apicid  : 31
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat pku ospke
vmx flags       : vnmi preemption_timer posted_intr invvpid ept_x_only ept_ad ept_1gb flexpriority apicv tsc_offset vtpr mtf vapic ept vpid unrestricted_guest vapic_reg vid pml
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa itlb_multihit
bogomips        : 3203.89
clflush size    : 64
cache_alignment : 64
address sizes   : 40 bits physical, 48 bits virtual
power management:

**Review by seemethere — APPROVED** [review]

**Comment by frank-wei** [comment]
@seemethere , I did some changes to the internal diff. Should I export it first?

**Comment by seemethere** [comment]
> @seemethere , I did some changes to the internal diff. Should I export it first?

Yes any change that you make to an internal diff that touches external code should be exported to Github. There is a land blocking job (that was bypassed) that will fail if that is not met, so I would pay attention to that signal next time since it will prevent things like this from happening.

There's no further action that needs to be done for this PR now, but in the future please export all changes to your Github PR.

**Comment by frank-wei** [comment]
> > @seemethere , I did some changes to the internal diff. Should I export it first?
> 
> Yes any change that you make to an internal diff that touches external code should be exported to Github. There is a land blocking job (that was bypassed) that will fail if that is not met, so I would pay attention to that signal next time since it will prevent things like this from happening.
> 
> There's no further action that needs to be done for this PR now, but in the future please export all changes to your Github PR.

Thanks. Sorry about the confusion.
